src/main/java/org/jabref/logic/openoffice/style/OOBibStyle.java

```diff
@@ -911,4 +958,392 @@ private String createAuthorList(String author, int maxAuthors, String andString,
+    }
+
+    protected LayoutFormatter getFieldFormatter() {
+        return fieldFormatter;
+    }
+
+    protected Map<EntryType, Layout> getBibLayout() {
+        return bibLayout;
+    }
+
+    protected Map<String, Object> getProperties() {
+        return properties;
+    }
+
+    protected Map<String, Object> getCitProperties() {
+        return citProperties;
+    }
+
+    protected void addJournal(String s) {
+        journals.add(s);
+    }
+
+    protected void setLocalCopy(String s) {
+        localCopy = s;
+    }
+
+    protected void setName(String s) {
+        name = s;
+    }
+
+    protected boolean getIsDefaultLayoutPresent() {
+        return isDefaultLayoutPresent;
+    }
+
+    protected void setIsDefaultLayoutPresent(boolean b) {
+        isDefaultLayoutPresent = b;
+    }
+
+    protected void setValid(boolean b) {
+        valid = b;
+    }
+
+    protected LayoutFormatterPreferences getPrefs() {
+        return prefs;
+    }
+
+    protected void setDefaultBibLayout(Layout l) {
+        defaultBibLayout = l;
+    }
+
+    /**
+     * Format a number-based citation marker for the given entries.
+     *
+     * @return The text for the citation.
+     */
+    public OOText getNumCitationMarker2(List<CitationMarkerNumericEntry> entries) {
+        final int minGroupingCount = this.getMinimumGroupingCount();
+        return OOBibStyleGetNumCitationMarker.getNumCitationMarker2(this,
+                                                                    entries,
+                                                                    minGroupingCount);
+    }
+
+    /**
+     * For some tests we need to override minGroupingCount.
+     */
+    public OOText getNumCitationMarker2(List<CitationMarkerNumericEntry> entries,
+                                        int minGroupingCount) {
+        return OOBibStyleGetNumCitationMarker.getNumCitationMarker2(this,
+                                                                    entries,
+                                                                    minGroupingCount);
+    }
+
+    /**
+     * Format a number-based bibliography label for the given number.
+     */
+    public OOText getNumCitationMarkerForBibliography(CitationMarkerNumericBibEntry entry) {
+        return OOBibStyleGetNumCitationMarker.getNumCitationMarkerForBibliography(this, entry);
+    }
+
+    /**
+     *  Make sure that (1) we have exactly one entry for each
+     *  citation, (2) each entry is either Optional.empty or its content is not empty when trimmed.
+     *
+     *  As a special case: pageInfos may be null. In this case
+     *  the result is a list filled with Optional.empty() values.
+     */
+    static List<Optional<OOText>>
+    normalizePageInfos(List<Optional<OOText>> pageInfos, int nCitations) {
+
+        // translate null to all-empty
+        if (pageInfos == null) {
+            List<Optional<OOText>> res = new ArrayList<>(nCitations);
```

Comment:
`normalizePageInfos` is not used anymore, removing.

Similar representation is still present elsewhere. So the questions remain.

> Is it really necessary to fill the list with Optional.empty? This looks a bit odd.

It does look odd. The original API (e.g. [`getCitationMarkers`](https://github.com/antalk2/jabref/blob/835e8f1bcc38316deb781d6cffc28c1ccb5641d6/src/main/java/org/jabref/logic/openoffice/style/OOBibStyle.java#L520)) used arrays and null values or entries.
Although originally there was only a single pageInfo for a citation group, now
we allow a pageInfo for each citation, and need a list or array to represent them.
When I needed to pass in multiple pageInfo values, I also used null values or null entries. 

[At this point](https://github.com/antalk2/jabref/blob/3db3080ee12ae7db9e8f8c4d5f6f626797a65f49/src/main/java/org/jabref/logic/openoffice/OOBibStyleGetCitationMarker.java#L499) normalizePageInfos ([OOBibStyle.regularizePageInfosForCitations](https://github.com/antalk2/jabref/blob/3db3080ee12ae7db9e8f8c4d5f6f626797a65f49/src/main/java/org/jabref/logic/openoffice/OOBibStyle.java#L331)) was used by functions accepting `null` for "no pageInfo".


Now we pass `List<CitationMarkerEntry>` and similar for citation marker generation, with entries providing `getPageInfo()`

After changing `getPageInfo()` to return Optional, it became confusing to convert its Optional.empty result to null
when collecting, and then back to Optional for the parts expecting one.
So I attempted to standardize on always using Optional for pageInfo values internally.

At the boundaries we are still receiving null, and need to consider `OODataModel.JabRef52` which
uses a single pageInfo per citation group (not per citation).

Where do we still use lists or arrays of pageInfo values?

- When creating a new citation group from the GUI,
we receive a single pageInfo, that is converted to `List<Optional<OOText>>`  in
[`insertCitationGroup`](https://github.com/antalk2/jabref/blob/835e8f1bcc38316deb781d6cffc28c1ccb5641d6/src/main/java/org/jabref/logic/openoffice/action/EditInsert.java#L77)  using [`OODataModel.fakePageInfos`](https://github.com/antalk2/jabref/blob/835e8f1bcc38316deb781d6cffc28c1ccb5641d6/src/main/java/org/jabref/model/openoffice/style/OODataModel.java#L22). This list is passed to `UpdateCitationMarkers.createAndFillCitationGroup`

- [`OODataModel.fakePageInfos`](https://github.com/antalk2/jabref/blob/835e8f1bcc38316deb781d6cffc28c1ccb5641d6/src/main/java/org/jabref/model/openoffice/style/OODataModel.java#L22)  is called in `Backend52.combinePageInfosCommon` for OODataModel.JabRef52 (to support `EditMerge.mergeCitationGroups`) (and in `EditInsert.insertCitationGroup` above)

- [`UpdateCitationMarkers.createAndFillCitationGroup`](https://github.com/antalk2/jabref/blob/835e8f1bcc38316deb781d6cffc28c1ccb5641d6/src/main/java/org/jabref/logic/openoffice/frontend/UpdateCitationMarkers.java#L122) is called in :
`EditInsert.insertCitationGroup` (above), `EditMerge.mergeCitationGroups` and `EditSeparate.separateCitations`.
The latter two are concerned about how to handle multiple pageInfo values  during a merge  and where to put the single pageInfo when breaking up a citation group (these are only problematic with the old datamodel). 

- `CitationEntry` still corresponds to a citation group (not a citation) (and allows a single pageInfo)
- In the document we still use `OODataModel.JabRef52`

- `OOFrontend.createCitationGroup` : called from `UpdateCitationMarkers.createAndFillCitationGroup`
- `Backend52.createCitationGroup` (`OODataModel.JabRef52` compatibility) called from `OOFrontend.createCitationGroup`

- `OOBibStyleTestHelper.getCitationMarker2` has parameter `String[] pageInfo` to provide API similar to
the old one, for easy comparison.
 
**summary**

The new API prefers citations to provide their pageInfo. We are passing around lists or arrays of pageInfo
values to handle the old datamodel. This includes the backend, EditMerge.mergeCitationGroups and EditSeparate.separateCitations, as well as interfacing to GUI parts (insertCitation and CitationEntry for ManageCitationsDialog). 

We could use null entries instead of Optional, as we did for some time. At least in some cases it is going to
involve Optional -> null on collection and null -> Optional when separating them.
Or storing null values and converting them in getPageInfo to Optional, and back to null when collecting.


> Can't you just skip them in a later step if there is no entry?

I am not sure I understand this. When is `later`? 
For `normalizePageInfos` it is history, but
[`OODataModel.fakePageInfos`](https://github.com/antalk2/jabref/blob/835e8f1bcc38316deb781d6cffc28c1ccb5641d6/src/main/java/org/jabref/model/openoffice/style/OODataModel.java#L22) is still in use, and produces
a `List<Optional<OOText>>` of normalized pageInfo values with a single (at most one) real (non-empty) entry.

It could return null or an empty list if the single entry becomes Optional.empty() itself.
Then at the locations using the result we should check and handle the special case. 
This is what I was trying to avoid by moving the special cases to OODataModel and Backend.

Where should we handle specially

- insertCitationGroup calls fakePageInfo, uses the result to generate a preliminary citation marker [here](https://github.com/antalk2/jabref/blob/835e8f1bcc38316deb781d6cffc28c1ccb5641d6/src/main/java/org/jabref/logic/openoffice/action/EditInsert.java#L83) and also passes it  to UpdateCitationMarkers.createAndFillCitationGroup, and ends up [Backend52.createCitationGroup](https://github.com/antalk2/jabref/blob/835e8f1bcc38316deb781d6cffc28c1ccb5641d6/src/main/java/org/jabref/logic/openoffice/backend/Backend52.java#L178)
- Backend52.combinePageInfosCommon calls fakePageInfos, called from Backend52.combinePageInfos which is called from mergeCitationGroups and the result is passed to UpdateCitationMarkers.createAndFillCitationGroup, and ends up in 
Backend52.createCitationGroup (same as above) 

This is two places, plus documentation on UpdateCitationMarkers.createAndFillCitationGroup, OOFrontend.createCitationGroup and Backend52.createCitationGroup stating that for Backend52 (or OODataModel.JabRef52) null is an accepted value.  I would suggest to wait with this. 

- Backend60 would either also have to handle the special case, or we will need to ensure that it does not receive
an abridged input. 
- If we do not create / use OODataModel.JabRef52 documents, we will not need the OODataModel.JabRef52 branch in
BackendXX.createCitationGroup and if we do not work with them, Backend.combinePageInfosCommon is not needed either.
- If [Cite special] in the gui learns to handle a pageInfo for each selected BibEntry, we will not need to go from a single
nullable entry to a list in insertCitationGroup